src/Common/OpenTelemetryTraceContext.h

```diff
@@ -19,6 +53,94 @@ struct OpenTelemetryTraceContext
     // Parse/compose OpenTelemetry traceparent header.
     bool parseTraceparentHeader(const std::string & traceparent, std::string & error);
     std::string composeTraceparentHeader() const;
+
+    bool isTraceEnabled() const
+    {
+        return trace_id != UUID();
+    }
+};
+
+/// tracing context kept on thread local
+struct OpenTelemetryThreadTraceContext : OpenTelemetryTraceContext
+{
+    OpenTelemetryThreadTraceContext& operator =(const OpenTelemetryTraceContext& context)
+    {
+        *(static_cast<OpenTelemetryTraceContext*>(this)) = context;
+        return *this;
+    }
+
+    void reset();
+
+    static const OpenTelemetryThreadTraceContext& current();
+
+    std::weak_ptr<OpenTelemetrySpanLog> span_log;
+};
+
+/// A scoped tracing context, is used to hold the tracing context at the beginning of each thread execution and clear the context automatically when the scope exists.
+/// It should be the root of all span logs for one tracing.
+///
+/// It's SAFE to construct this object multiple times on one same thread,
+/// but it's not encourage to do so because this is only a protection in case of code changes.
```

Comment:
Let me be clear. 

This context should be initialized once at the start of processing on each thread by constructing the object in the code somewhere, let's assume it's at `MethodA`. 

But if one day there're some refactoring on these code, and `MethodA` is no longer the start entry but `MethodB` becomes the start, and `MethodB` initializes the context without realizing the `MethodA` contains the code to initialize the context.

Now, on the same thread, the context is initialized twice. If there's no such protection, spans generated after `MethodA` and `MethodB` have wrong parent, they look like this:

```
Parent
    |-------Span1 ( generated after MethodB )
    |-------Span2 ( generated after MethodA )
```

But actually Span2 should be child of Span1.

Another problem is the context restoring in the destructor of this object. Since the context on thread local has been initialized twice, when `MethodA` exits to destruct the context, the whole context on thread local destroys,  spans generated after this phase can't be recorded.